// runnable/lib/api-client.js
'use strict';

var util = require('util');
var ApiClient = require('simple-api-client');
var findIndex = require('101/find-index');
var isFunction = require('101/is-function');
var isObject = require('101/is-object');
var UnexpectedError = require('./unexpected-error');
var Boom = require('boom');

module.exports = Base;

function Base (host, opts) {
  if (isObject(host)) {
    opts = host;
    host = null;
  }
  this.host = host || 'http://api.runnable.com';
  this.opts = opts || {};
  ApiClient.apply(this, arguments);
}

util.inherits(Base, ApiClient);

// overwrite methods to handle opts.statusCodes
// Base.prototype.{post, get, put, patch, del, ...}
var Super = ApiClient.prototype;
var nonHttpMethods = ['request', 'xhr'];
Object.keys(ApiClient.prototype).forEach(function (method) {
  if (~nonHttpMethods.indexOf(method)) { return; }
  Base.prototype[method] = function (/* path[, opts][, cb] */) {
    var args = Array.prototype.slice.call(arguments);
    var optsIndex = findIndex(args, isObject);
    var cbIndex = findIndex(args, isFunction);
    var opts;
    // default opts
    if (~optsIndex) {
      opts = args[optsIndex];
    }
    else if (~cbIndex) {
      opts = {};
      args.splice(cbIndex, 0, opts);
    }
    else {
      opts = {};
      args.push(opts);
    }
    // handleErrors
    if (~cbIndex) {
      var cb = args[cbIndex];
      args[cbIndex] = this.handleErrors(method, args[0], opts, cb);
      cb = args[cbIndex];
    }
    // attach runnable token
    if (this.opts.token) {
      opts.headers = {
        'runnable-token': this.opts.token
      };
    }
    // call super method
    return Super[method].apply(this, args);
  };
});

Base.prototype.handleErrors = function (method, path, opts, cb) {
  var self = this;
  var statusCodes = opts.statusCodes;
  return function (err, res, body) {
    if (err) {
      cbErr(err);
    }
    else if (statusCodes) {
      var code = res.statusCode;
      if (statusCodes[code] === true) { // success
        cb(err, body, code);
      }
      else if (code in statusCodes) {
        var message = statusCodes[code] || body.message || body;
        err = Boom.create(code, message, body);
        cbErr(err);
      }
      else {
        err = UnexpectedError.create(res, body);
        cbErr(err);
      }
    }
    else {
      cb(null, res, body);
    }
  };
  function cbErr (err) {
    err.url = {
      host: self.host,
      method: method,
      path: path
    };
    cb(err);
  }
};
